sysdump/sysdump.go

```diff
@@ -0,0 +1,815 @@
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package sysdump
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"os"
+	"path"
+	"path/filepath"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/cilium/cilium-cli/internal/utils"
+	"github.com/cilium/workerpool"
+	"github.com/hashicorp/go-multierror"
+	archiver "github.com/mholt/archiver/v3"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/cli-runtime/pkg/genericclioptions"
+)
+
+type SysdumpOptions struct {
+	CiliumLabelSelector         string
+	CiliumNamespace             string
+	CiliumOperatorLabelSelector string
+	CiliumOperatorNamespace     string
+	Debug                       bool
+	HubbleLabelSelector         string
+	HubbleNamespace             string
+	HubbleRelayLabelSelector    string
+	HubbleRelayNamespace        string
+	HubbleUILabelSelector       string
+	HubbleUINamespace           string
+	LargeSysdumpAbortTimeout    time.Duration
+	LargeSysdumpThreshold       int32
+	LogsLimitBytes              int64
+	LogsSinceTime               time.Duration
+	NodeList                    string
+	OutputFileName              string
+	Quick                       bool
+	RESTClientGetter            genericclioptions.RESTClientGetter
+	WorkerCount                 int32
+	Writer                      io.Writer
+}
+
+type Sysdump interface {
+	Collect() error
+}
+
+type sysdumpTask func(context.Context) error
+
+type sysdump struct {
+	client  KubernetesClient
+	options SysdumpOptions
+	pool    *workerpool.WorkerPool
+	// tasksWg is used to wait for sub-tasks to be submitted to the pool before calling 'Drain'.
+	// It is required since we don't know beforehand how many sub-tasks will be created, as they depend on the number of Cilium/Hubble/... pods found by "main" tasks.
+	// Main tasks should defer taskWg.Done() and proceed to submitting sub-tasks.
+	tasksWg sync.WaitGroup
+}
+
+func NewSysdump(k KubernetesClient, o SysdumpOptions) Sysdump {
+	return &sysdump{
+		client:  k,
+		options: o,
+		pool:    workerpool.New(int(o.WorkerCount)),
+	}
+}
+
+func (s *sysdump) Collect() error {
+	// Grab the current timestamp and create a temporary directory to hold the files.
+	t := time.Now()
+
+	// replaceTimestamp can be used to replace the special timestamp placeholder in file and directory names.
+	replaceTimestamp := func(f string) string {
+		return strings.Replace(f, timestampPlaceholderFileName, t.Format(timeFormat), -1)
+	}
+
+	d, err := ioutil.TempDir("", "*")
```

Comment:
use `os.TempDir`

Reply:
Hmmm I think [`os.TempDir`](https://golang.org/pkg/os/#TempDir) is a slightly different thing — it would just return, say, `/tmp` (or something weird in MacOS). We'd still have to create a directory there. [`ioutil.TempDir`](https://golang.org/pkg/io/ioutil/#TempDir) actually creates a directory inside the temporary directory, so it's one step instead of two. WDYT?

Reply:
The whole `ioutil` package [is deprecated](https://golang.org/doc/go1.16#ioutil) since the release of Go v1.16. Instead of `ioutil.TempDir`, [`os.MkdirTemp`](https://golang.org/pkg/os/#MkdirTemp) should be used instead.